File: Medicine_business/medicine-common/medicine-common-redis/src/main/java/com/zhangyichuang/medicine/common/redis/core/RedisHashCache.java
package com.zhangyichuang.medicine.common.redis.core;

import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.data.redis.core.Cursor;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.data.redis.core.ScanOptions;
import org.springframework.stereotype.Component;
import org.springframework.util.StringUtils;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.TimeUnit;

/**
 * Redis Hash 操作缓存工具类
 * 提供 Hash 数据结构的各种操作方法，包括批量扫描功能
 *
 * @author Chuang
 */
@Component
@RequiredArgsConstructor
@SuppressWarnings(value = {"unchecked", "rawtypes"})
public final class RedisHashCache {

    /**
     * 日志记录器。
     */
    private static final Logger log = LoggerFactory.getLogger(RedisHashCache.class);

    private final static int scanCount = 1000;
    private final RedisTemplate redisTemplate;

    /**
     * 向 Hash 添加单个字段
     *
     * @param key   Redis 键
     * @param field 字段名
     * @param value 字段值
     * @param <HK>  字段类型
     * @param <HV>  值类型
     */
    public <HK, HV> void hPut(final String key, final HK field, final HV value) {
        redisTemplate.opsForHash().put(key, field, value);
    }

    /**
     * 向 Hash 批量添加字段
     *
     * @param key      Redis 键
     * @param fieldMap 字段–值 对
     * @param <HK>     字段类型
     * @param <HV>     值类型
     */
    public <HK, HV> void hPutAll(final String key, final Map<HK, HV> fieldMap) {
        redisTemplate.opsForHash().putAll(key, fieldMap);
    }

    /**
     * 从 Hash 中取单个字段
     *
     * @param key   Redis 键
     * @param field 字段名
     * @param <HK>  字段类型
     * @param <HV>  值类型
     * @return 对应字段的值；不存在返回 null
     */
    @SuppressWarnings("unchecked")
    public <HK, HV> HV hGet(final String key, final HK field) {
        return (HV) redisTemplate.opsForHash().get(key, field);
    }

    /**
     * 取整个 Hash 的所有字段和值
     *
     * @param key  Redis 键
     * @param <HK> 字段类型
     * @param <HV> 值类型
     * @return 一个 Map，key 对应字段，value 对应值
     */
    @SuppressWarnings("unchecked")
    public <HK, HV> Map<HK, HV> hGetAll(final String key) {
        return (Map<HK, HV>) redisTemplate.opsForHash().entries(key);
    }


    /**
     * 从 Hash 中删除一个或多个字段
     *
     * @param key    Redis 键
     * @param fields 要删除的字段名
     * @param <HK>   字段类型
     * @return 删除的字段数量
     */
    public <HK> Long hRemove(final String key, final HK... fields) {
        return redisTemplate.opsForHash().delete(key, (Object[]) fields);
    }

    /**
     * 判断 Hash 中是否存在某个字段
     *
     * @param key   Redis 键
     * @param field 字段名
     * @return true=存在；false=不存在
     */
    public Boolean hExists(final String key, final Object field) {
        return redisTemplate.opsForHash().hasKey(key, field);
    }

    /**
     * 获取 Hash 中字段的数量
     *
     * @param key Redis 键
     * @return 字段个数
     */
    public Long hSize(final String key) {
        return redisTemplate.opsForHash().size(key);
    }

    /**
     * 对 Hash 中的数值字段做增量操作
     *
     * @param key   Redis 键
     * @param field 字段名
     * @param delta 增量（负值表示减）
     * @param <HK>  字段类型
     * @return 操作后的新值
     */
    public <HK> Long hIncrement(final String key, final HK field, final long delta) {
        return redisTemplate.opsForHash().increment(key, field, delta);
    }

    /**
     * 给任意 key 设置过期时间（Hash 也能用）
     *
     * @param key      Redis 键
     * @param timeout  时长
     * @param timeUnit 时间单位
     * @return 设置是否成功
     */
    public Boolean expire(final String key, final long timeout, final TimeUnit timeUnit) {
        return redisTemplate.expire(key, timeout, timeUnit);
    }

    /**
     * 查询 key 的剩余存活秒数
     *
     * @param key Redis 键
     * @return 剩余秒数；-2=不存在，-1=未设置过期
     */
    public Long getExpire(final String key) {
        return redisTemplate.getExpire(key, TimeUnit.SECONDS);
    }

    /**
     * 批量键扫描方法 - 使用 Redis SCAN 操作高效获取匹配的 Hash 键
     * 推荐在生产环境中使用此方法，避免阻塞 Redis 服务器
     *
     * @param keyPattern Redis 键模式，支持通配符（如：auth:session:index:*）
     * @return 匹配的键列表，如果没有匹配的键则返回空列表
     */
    public List<String> scanKeys(final String keyPattern) {
        if (!StringUtils.hasText(keyPattern)) {
            log.warn("Redis哈希扫描模式为空，返回空列表");
            return new ArrayList<>();
        }

        List<String> keys = new ArrayList<>();
        try {
            ScanOptions options = ScanOptions.scanOptions()
                    .match(keyPattern)
                    .count(scanCount)
                    .build();

            try (Cursor<String> cursor = redisTemplate.scan(options)) {
                while (cursor.hasNext()) {
                    keys.add(cursor.next());
                }
            }

        } catch (Exception e) {
            log.error("Redis Hash scan keys failed, pattern: {}, error: {}", keyPattern, e.getMessage(), e);
            // 发生异常时返回空列表，避免影响业务流程
            return new ArrayList<>();
        }

        return keys;
    }

    /**
     * 批量值获取方法 - 获取匹配模式的 Hash 键及其所有字段值
     * 使用 Redis SCAN 操作高效扫描键，然后批量获取每个 Hash 的所有字段值
     *
     * @param keyPattern Redis 键模式，支持通配符（如：auth:session:index:*）
     * @return 嵌套Map，外层键为Hash键，内层Map为该Hash的所有字段值，如果没有匹配的键则返回空Map
     */
    public Map<String, Map<Object, Object>> scanKeysWithValues(final String keyPattern) {
        if (!StringUtils.hasText(keyPattern)) {
            return new HashMap<>();
        }

        // 首先扫描获取所有匹配的键
        List<String> keys = scanKeys(keyPattern);
        if (keys.isEmpty()) {
            return new HashMap<>();
        }

        Map<String, Map<Object, Object>> result = new HashMap<>();
        try {
            // 逐个获取每个 Hash 的所有字段值
            for (String key : keys) {
                Map<Object, Object> hashData = redisTemplate.opsForHash().entries(key);
                result.put(key, hashData);
            }
        } catch (Exception e) {
            log.error("Redis Hash scan keys with values failed, pattern: {}, error: {}", keyPattern, e.getMessage(), e);
            return new HashMap<>();
        }

        return result;
    }


}

File: medicine-ai-agent/app/core/database/redis/redis_hash_cache.py
from __future__ import annotations

from typing import Any

from redis import Redis
from redis.exceptions import RedisError

from app.core.database.redis.config import get_redis_connection
from app.core.database.redis.redis_cache import (
    DEFAULT_SCAN_COUNT,
    _decode_key,
    _raise_redis_operation_error,
)


class RedisHashCache:
    """Redis Hash 结构操作封装。"""

    def __init__(self, redis_client: Redis | None = None) -> None:
        """初始化 Redis Hash 操作封装。

        Args:
            redis_client: 可选 Redis 客户端，未传时使用全局连接。
        """
        self._redis = redis_client or get_redis_connection()

    def h_put(self, key: str, field: Any, value: Any) -> None:
        """写入单个 Hash 字段。

        Args:
            key: Redis 键。
            field: Hash 字段名。
            value: 字段值。

        Raises:
            ServiceException: Redis 写入失败时抛出。
        """
        try:
            self._redis.hset(key, field, value)
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hset", exc=exc) from exc

    def h_put_all(self, key: str, field_map: dict[Any, Any]) -> None:
        """批量写入 Hash 字段。

        Args:
            key: Redis 键。
            field_map: 字段映射。

        Raises:
            ServiceException: Redis 写入失败时抛出。
        """
        try:
            self._redis.hset(key, mapping=field_map)
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hset_many", exc=exc) from exc

    def h_get(self, key: str, field: Any) -> Any:
        """读取单个 Hash 字段。

        Args:
            key: Redis 键。
            field: Hash 字段名。

        Returns:
            Any: 字段值，不存在时返回 `None`。

        Raises:
            ServiceException: Redis 读取失败时抛出。
        """
        try:
            return self._redis.hget(key, field)
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hget", exc=exc) from exc

    def h_get_all(self, key: str) -> dict[Any, Any]:
        """读取整个 Hash。

        Args:
            key: Redis 键。

        Returns:
            dict[Any, Any]: 全量字段映射。

        Raises:
            ServiceException: Redis 读取失败时抛出。
        """
        try:
            return dict(self._redis.hgetall(key))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hgetall", exc=exc) from exc

    def h_remove(self, key: str, *fields: Any) -> int:
        """删除一个或多个 Hash 字段。

        Args:
            key: Redis 键。
            *fields: 要删除的字段列表。

        Returns:
            int: 实际删除的字段数。

        Raises:
            ServiceException: Redis 删除失败时抛出。
        """
        if not fields:
            return 0
        try:
            return int(self._redis.hdel(key, *fields))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hdel", exc=exc) from exc

    def h_exists(self, key: str, field: Any) -> bool:
        """判断 Hash 字段是否存在。

        Args:
            key: Redis 键。
            field: Hash 字段名。

        Returns:
            bool: 存在返回 `True`，否则返回 `False`。

        Raises:
            ServiceException: Redis 查询失败时抛出。
        """
        try:
            return bool(self._redis.hexists(key, field))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hexists", exc=exc) from exc

    def h_size(self, key: str) -> int:
        """读取 Hash 字段数。

        Args:
            key: Redis 键。

        Returns:
            int: 字段总数。

        Raises:
            ServiceException: Redis 查询失败时抛出。
        """
        try:
            return int(self._redis.hlen(key))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hlen", exc=exc) from exc

    def h_increment(self, key: str, field: Any, delta: int) -> int:
        """对 Hash 数值字段做增量。

        Args:
            key: Redis 键。
            field: Hash 字段名。
            delta: 增量值。

        Returns:
            int: 增量后的字段值。

        Raises:
            ServiceException: Redis 自增失败时抛出。
        """
        try:
            return int(self._redis.hincrby(key, field, delta))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="hincrby", exc=exc) from exc

    def expire(self, key: str, timeout: int) -> bool:
        """设置键过期时间（秒）。

        Args:
            key: Redis 键。
            timeout: 过期秒数。

        Returns:
            bool: 设置成功返回 `True`，否则返回 `False`。

        Raises:
            ServiceException: Redis 设置过期失败时抛出。
        """
        try:
            return bool(self._redis.expire(key, timeout))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="expire", exc=exc) from exc

    def get_expire(self, key: str) -> int:
        """读取键剩余过期时间（秒）。

        Args:
            key: Redis 键。

        Returns:
            int: 剩余秒数；`-1` 表示未设置过期，`-2` 表示键不存在。

        Raises:
            ServiceException: Redis 查询 TTL 失败时抛出。
        """
        try:
            return int(self._redis.ttl(key))
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="ttl", exc=exc) from exc

    def scan_keys(self, key_pattern: str, *, count: int = DEFAULT_SCAN_COUNT) -> list[str]:
        """按模式扫描 Hash 键列表。

        Args:
            key_pattern: 键匹配模式。
            count: 每次 SCAN 的建议批次大小。

        Returns:
            list[str]: 匹配到的键列表。

        Raises:
            ServiceException: Redis 扫描失败时抛出。
        """
        cursor = 0
        keys: list[str] = []
        try:
            while True:
                cursor, batch = self._redis.scan(
                    cursor=cursor,
                    match=key_pattern,
                    count=count,
                )
                keys.extend(_decode_key(item) for item in batch)
                if int(cursor) == 0:
                    break
            return keys
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="scan", exc=exc) from exc

    def scan_keys_with_values(
            self,
            key_pattern: str,
            *,
            count: int = DEFAULT_SCAN_COUNT,
    ) -> dict[str, dict[Any, Any]]:
        """按模式扫描 Hash 键并读取每个 Hash 全量字段。

        Args:
            key_pattern: 键匹配模式。
            count: 每次 SCAN 的建议批次大小。

        Returns:
            dict[str, dict[Any, Any]]: 键到 Hash 字段映射的映射。

        Raises:
            ServiceException: Redis 扫描或读取失败时抛出。
        """
        matched_keys = self.scan_keys(key_pattern, count=count)
        if not matched_keys:
            return {}
        result: dict[str, dict[Any, Any]] = {}
        try:
            for key in matched_keys:
                result[key] = dict(self._redis.hgetall(key))
            return result
        except RedisError as exc:
            raise _raise_redis_operation_error(operation="scan_hash_values", exc=exc) from exc
